pkg/user/parallel_tx_submission.go

```diff
@@ -0,0 +1,528 @@
+package user
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"cosmossdk.io/x/feegrant"
+	"github.com/celestiaorg/celestia-app/v6/pkg/appconsts"
+	"github.com/celestiaorg/go-square/v3/share"
+	"github.com/cosmos/cosmos-sdk/crypto/hd"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	bank "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+// SubmissionJob represents a transaction submission task for parallel processing
+type SubmissionJob struct {
+	ID       string
+	Blobs    []*share.Blob
+	Options  []TxOption
+	Ctx      context.Context
+	ResultsC chan SubmissionResult
+}
+
+// SubmissionResult contains the result of a parallel transaction submission
+type SubmissionResult struct {
+	Signer     string
+	TxResponse *TxResponse
+	Error      error
+}
+
+// ParallelTxPool manages parallel transaction submission
+type ParallelTxPool struct {
+	mtx         sync.RWMutex
+	client      *TxClient
+	jobQueue    chan *SubmissionJob
+	workers     []*TxWorker
+	started     atomic.Bool
+	ctx         context.Context
+	cancel      context.CancelFunc
+	initialize  bool        // whether to initialize workers automatically
+	initialized atomic.Bool // whether workers have been initialized
+}
+
+// TxWorker represents a worker that processes transactions using a specific account
+type TxWorker struct {
+	id          int
+	accountName string
+	address     string
+	client      *TxClient
+	jobQueue    chan *SubmissionJob
+}
+
+const (
+	defaultParallelQueueSize = 100
+)
+
+func newParallelTxPool(client *TxClient, numWorkers int, initialize bool) *ParallelTxPool {
+	pool := &ParallelTxPool{
+		client:     client,
+		jobQueue:   make(chan *SubmissionJob, defaultParallelQueueSize),
+		workers:    make([]*TxWorker, numWorkers),
+		initialize: initialize,
+	}
+
+	// Create workers: first worker always uses existing signer account
+	for i := 0; i < numWorkers; i++ {
+		var accountName, address string
+
+		if i == 0 {
+			// First worker uses the existing default account
+			accountName = client.DefaultAccountName()
+			address = client.DefaultAddress().String()
+		} else {
+			// Additional workers use generated account names
+			accountName = fmt.Sprintf("parallel-worker-%d", i)
+
+			// Get worker address from keyring if account exists
+			if record, err := client.signer.keys.Key(accountName); err == nil {
+				if addr, err := record.GetAddress(); err == nil {
+					address = addr.String()
+				}
+			}
+		}
+
+		worker := &TxWorker{
+			id:          i,
+			accountName: accountName,
+			address:     address,
+			client:      client,
+			jobQueue:    pool.jobQueue,
+		}
+		pool.workers[i] = worker
+	}
+
+	return pool
+}
+
+// NewParallelTxPool creates a new parallel transaction submission pool.
+func NewParallelTxPool(client *TxClient, numWorkers int, initialize bool) *ParallelTxPool {
+	return newParallelTxPool(client, numWorkers, initialize)
+}
```

Comment:
[silly question] this public method seems like a no-op wrapper around the private method. Could we eliminate these lines and make the private method public?

Reply:
https://github.com/celestiaorg/celestia-app/pull/5776/commits/4c02643ce5f19a279ee00bd2f667c1fc971fb3f9